File: ts/src/messages.ts
/**
 * The spec's shared message-template set (see conformance/golden/
 * messages.json — written by the Go reference; this table must match it,
 * the conformance suite compares them). Placeholders {expected} and {actual}
 * substitute display forms; a missing value renders "…". RULE_VIOLATED /
 * EXPR_ERROR / INVALID_SCHEMA carry free text and are not templated.
 */

import { ErrorCode } from "./gen/schemapb/errors_pb.js";
import type { Value } from "./gen/schemapb/value_pb.js";
import { displayString } from "./render.js";
import { toNative } from "./value.js";

export const messageTemplates: ReadonlyMap<ErrorCode, string> = new Map([
  [ErrorCode.TYPE_MISMATCH, "expected {expected}"],
  [ErrorCode.REQUIRED_MISSING, "required"],
  [ErrorCode.NOT_NULLABLE, "must not be null"],
  [ErrorCode.UNKNOWN_FIELD, "unknown field"],
  [ErrorCode.IMMUTABLE_MODIFIED, "immutable: cannot be changed"],

  [ErrorCode.CONST_MISMATCH, "must equal {expected}"],
  [ErrorCode.GT_VIOLATED, "must be > {expected}"],
  [ErrorCode.GTE_VIOLATED, "must be >= {expected}"],
  [ErrorCode.LT_VIOLATED, "must be < {expected}"],
  [ErrorCode.LTE_VIOLATED, "must be <= {expected}"],
  [ErrorCode.NOT_IN_ALLOWED_SET, "must be one of {expected}"],
  [ErrorCode.IN_FORBIDDEN_SET, "must not be one of {expected}"],
  [ErrorCode.MULTIPLE_OF_VIOLATED, "must be a multiple of {expected}"],

  [ErrorCode.LEN_MISMATCH, "length must be exactly {expected}"],
  [ErrorCode.MIN_LEN_VIOLATED, "length must be at least {expected}"],
  [ErrorCode.MAX_LEN_VIOLATED, "length must be at most {expected}"],
  [ErrorCode.PATTERN_MISMATCH, "must match pattern {expected}"],
  [ErrorCode.FORMAT_MISMATCH, "must be a valid {expected}"],
  [ErrorCode.PREFIX_MISMATCH, "must start with {expected}"],
  [ErrorCode.SUFFIX_MISMATCH, "must end with {expected}"],
  [ErrorCode.UNSUPPORTED_FORMAT, "format {expected} is not supported by this implementation"],

  [ErrorCode.CHOICE_NOT_ALLOWED, "must be one of {expected}"],

  [ErrorCode.MIN_ITEMS_VIOLATED, "must have at least {expected} items"],
  [ErrorCode.MAX_ITEMS_VIOLATED, "must have at most {expected} items"],
  [ErrorCode.NOT_UNIQUE, "must be unique"],
  [ErrorCode.LIST_COUNT_MISMATCH, "must have exactly {expected} items"],

  [ErrorCode.MIN_ENTRIES_VIOLATED, "must have at least {expected} entries"],
  [ErrorCode.MAX_ENTRIES_VIOLATED, "must have at most {expected} entries"],

  [ErrorCode.MIN_PROPERTIES_VIOLATED, "must have at least {expected} properties"],
  [ErrorCode.MAX_PROPERTIES_VIOLATED, "must have at most {expected} properties"],

  [ErrorCode.DISCRIMINATOR_MISSING, "discriminator {expected} must be a non-empty string"],
  [ErrorCode.UNKNOWN_VARIANT, "unknown variant {actual}"],
  [ErrorCode.UNKNOWN_REF, "unknown $ref {expected}"],
]);

/** Renders a code's template with the typed values. */
export function renderMessage(
  code: ErrorCode,
  expected: Value | undefined,
  actual: Value | undefined,
): string {
  const tpl = messageTemplates.get(code);
  if (tpl === undefined) {
    return "";
  }
  const sub = (s: string, ph: string, v: Value | undefined): string => {
    if (!s.includes(ph)) {
      return s;
    }
    const disp = v === undefined ? "…" : displayString(toNative(v));
    return s.replaceAll(ph, disp);
  };
  return sub(sub(tpl, "{expected}", expected), "{actual}", actual);
}

File: rust/src/messages.rs
//! The spec's shared message-template set (must match
//! conformance/golden/messages.json, written by the Go reference).

use crate::gen::schemapb::{ErrorCode, Value};
use crate::render::display_string;
use crate::value::to_native;

/// The template for one error code; `None` for the free-text codes
/// (`RuleViolated`, `ExprError`, `InvalidSchema`).
#[must_use]
pub const fn template(code: ErrorCode) -> Option<&'static str> {
    use ErrorCode as C;
    match code {
        C::TypeMismatch => Some("expected {expected}"),
        C::RequiredMissing => Some("required"),
        C::NotNullable => Some("must not be null"),
        C::UnknownField => Some("unknown field"),
        C::ImmutableModified => Some("immutable: cannot be changed"),
        C::ConstMismatch => Some("must equal {expected}"),
        C::GtViolated => Some("must be > {expected}"),
        C::GteViolated => Some("must be >= {expected}"),
        C::LtViolated => Some("must be < {expected}"),
        C::LteViolated => Some("must be <= {expected}"),
        C::NotInAllowedSet => Some("must be one of {expected}"),
        C::InForbiddenSet => Some("must not be one of {expected}"),
        C::MultipleOfViolated => Some("must be a multiple of {expected}"),
        C::LenMismatch => Some("length must be exactly {expected}"),
        C::MinLenViolated => Some("length must be at least {expected}"),
        C::MaxLenViolated => Some("length must be at most {expected}"),
        C::PatternMismatch => Some("must match pattern {expected}"),
        C::FormatMismatch => Some("must be a valid {expected}"),
        C::PrefixMismatch => Some("must start with {expected}"),
        C::SuffixMismatch => Some("must end with {expected}"),
        C::UnsupportedFormat => Some("format {expected} is not supported by this implementation"),
        C::ChoiceNotAllowed => Some("must be one of {expected}"),
        C::MinItemsViolated => Some("must have at least {expected} items"),
        C::MaxItemsViolated => Some("must have at most {expected} items"),
        C::NotUnique => Some("must be unique"),
        C::ListCountMismatch => Some("must have exactly {expected} items"),
        C::MinEntriesViolated => Some("must have at least {expected} entries"),
        C::MaxEntriesViolated => Some("must have at most {expected} entries"),
        C::MinPropertiesViolated => Some("must have at least {expected} properties"),
        C::MaxPropertiesViolated => Some("must have at most {expected} properties"),
        C::DiscriminatorMissing => Some("discriminator {expected} must be a non-empty string"),
        C::UnknownVariant => Some("unknown variant {actual}"),
        C::UnknownRef => Some("unknown $ref {expected}"),
        C::Unspecified | C::RuleViolated | C::ExprError | C::InvalidSchema => None,
    }
}

/// Renders a code's template with the typed values ("…" for absent ones).
#[must_use]
pub fn render_message(code: ErrorCode, expected: Option<&Value>, actual: Option<&Value>) -> String {
    let Some(tpl) = template(code) else {
        return String::new();
    };
    let sub = |s: &str, placeholder: &str, v: Option<&Value>| -> String {
        if !s.contains(placeholder) {
            return s.to_owned();
        }
        let disp = v.map_or_else(|| "…".to_owned(), |x| display_string(&to_native(Some(x))));
        s.replace(placeholder, &disp)
    };
    sub(&sub(tpl, "{expected}", expected), "{actual}", actual)
}
